### PR TITLE
Altinn AGA/FTR invoice sending to Fakturabank

### LINKED ISSUE #190: Send AGA and FTR from lodo to fakturabank
2016-06-16_14-59-20_send_altinn_to_fakturabank

### PR DESCRIPTION
**Pull request by mladjo2505**
Closes #190 
Might be some more things that need to be implemented/changed for this to work 100% as imagined.
New [kommuner.csv](https://github.com/dodoas/lodo/files/365364/kommuner.csv.zip) file.

_EDIT_ 2016-07-26 new file [kommuner.csv](https://github.com/dodoas/lodo/files/383277/kommuner.csv.zip) file.

### CODE REVIEW

**Comment by mladjo2505** [comment]
Based on kommune branch since some of the things implemented there were needed for this.
Should be rebased after #193 is merged.

**Comment by mladjo2505** [comment]
- We are missing buyer name.
- Send supplier contact information (where to get it from?)
- Do we send to FB NO:ORGNR for the sub-company or parent company? In the xml response we only get the parent company's NO:ORGNR
- Should this invoice be to only tax?

File: modules/kommune/model/kommune.class.php [hunk 1]
```diff
@@ -125,6 +128,26 @@ public function load($kommune_id) {
...
   }
 
+  // Sets the properties of this object to the columns from the database for column value
+  // Sets to the first occurance where all given columns are the same as the given ones
+  public function load_by_field_value($args) {
+    $matched_id = NULL;
+    foreach ($this->kommunes_in_db_by_id as $kommune_id => $kommune_object) {
+      $matches = true;
+      foreach ($args as $column_name => $column_value) {
+        if (in_array($column_name, $this->permitted_columns) && $kommune_object->{$column_name} != $args[$column_name]) {
```
Comment (mladjo2505) [inline modules/kommune/model/kommune.class.php]: Change `$args[$column_name]` to `$column_value`

File: modules/invoice/model/invoice.class.php [hunk 2]
```diff
@@ -1326,5 +1328,143 @@ function CheckIfAnythingChanged($InvoiceID, $InvoiceLineIDs, &$args) {
...
+      }
+      $issue_date = strftime('%F', $issue_date);
+      $bank_account_number = (string) $xml->Mottak->innbetalingsinformasjon->kontonummer;
+      $kommune = new kommune();
+      $kommune->load_by_field_value(array('BankAccountNumber' => $bank_account_number));
+      $kommune_orgno = preg_replace('/[^0-9]/', '', $kommune->OrgNumber);
+      $kommune_name = $kommune->OrgName;
+      $customer_orgno = (string) $xml->Mottak->innsender->norskIdentifikator;
+      $customer_orgno = preg_replace('/[^0-9]+/', '', $customer_orgno);
+      $subcompany_query = "SELECT * FROM subcompany WHERE OrgNumber = '$customer_orgno'";
```
Comment (mladjo2505) [inline modules/invoice/model/invoice.class.php]: Possibly change if we need to get the parent company and not sub-company.

File: modules/invoice/model/invoice.class.php [hunk 3]
```diff
@@ -1326,5 +1328,143 @@ function CheckIfAnythingChanged($InvoiceID, $InvoiceLineIDs, &$args) {
...
+        $kid = (string) $xml->Mottak->innbetalingsinformasjon->kidForArbeidsgiveravgift;
+        $amount = (float) $xml->Mottak->mottattPeriode->mottattAvgiftOgTrekkTotalt->sumArbeidsgiveravgift;
+      } else {
+        $kid = (string) $xml->Mottak->innbetalingsinformasjon->kidForForskuddstrekk;
+        $amount = (float) $xml->Mottak->mottattPeriode->mottattAvgiftOgTrekkTotalt->sumForskuddstrekk;
+      }
+
+      $altinn_invoice->ID = $invoice_type . '-' . $altinn_reference;
+      $altinn_invoice->IssueDate = $issue_date;
+      $altinn_invoice->Note = $invoice_type . ' ' . strftime('%Y-%m');
```
Comment (mladjo2505) [inline modules/invoice/model/invoice.class.php]: `strftime()` should be from `Mottak->kalendermaaned`

Comment (mladjo2505) [inline modules/invoice/model/invoice.class.php]: Do we need this Note node?

File: modules/invoice/model/invoice.class.php [hunk 4]
```diff
@@ -1326,5 +1328,143 @@ function CheckIfAnythingChanged($InvoiceID, $InvoiceLineIDs, &$args) {
...
+
+      // Invoice supplier
+      $altinn_invoice->AccountingSupplierParty->Party->PartyLegalEntity->CompanyID = $kommune_orgno;
+      $altinn_invoice->AccountingSupplierParty->Party->PartyTaxScheme->CompanyID = $kommune_orgno . ' MVA';
+      $altinn_invoice->AccountingSupplierParty->Party->PartyTaxScheme->CompanyIDSchemeID = 'NO:ORGNR';
+
+      $altinn_invoice->AccountingSupplierParty->Party->PartyName->Name = $kommune_name;
+/*
+      Missing info! Can't find a way to have the address and contact info.
+      Maybe add the kommune as a supplier and link it to a kommune or add more columns to kommune table.
```
Comment (mladjo2505) [inline modules/invoice/model/invoice.class.php]: What should we do about this?

File: modules/invoice/model/invoice.class.php [hunk 5]
```diff
@@ -1326,5 +1328,143 @@ function CheckIfAnythingChanged($InvoiceID, $InvoiceLineIDs, &$args) {
...
+      $altinn_invoice->AccountingSupplierParty->Party->Contact->ElectronicMail = '';
+
+ */
+
+      // Invoice customer
+      $altinn_invoice->AccountingCustomerParty->Party->PartyLegalEntity->CompanyID = $customer_orgno;
+      $altinn_invoice->AccountingCustomerParty->Party->PartyLegalEntity->CompanyIDSchemeID = 'NO:ORGNR';
+      $altinn_invoice->AccountingCustomerParty->Party->PartyName->Name = $customer_name;
+
+      // Missing info! Using company's address and contact info.
```
Comment (mladjo2505) [inline modules/invoice/model/invoice.class.php]: If the customer is the parent company this should be fine.

File: modules/invoice/model/invoice.class.php [hunk 6]
```diff
@@ -1326,5 +1328,143 @@ function CheckIfAnythingChanged($InvoiceID, $InvoiceLineIDs, &$args) {
...
+      $altinn_invoice->AccountingCustomerParty->Party->Contact->Mobile = $_lib['sess']->get_companydef('Mobile');
+      $altinn_invoice->AccountingCustomerParty->Party->Contact->Telefax = $_lib['sess']->get_companydef('Fax');
+      $altinn_invoice->AccountingCustomerParty->Party->Contact->ElectronicMail = $_lib['sess']->get_companydef('Email');
+
+      // Invoice payment means
+      $altinn_invoice->PaymentMeans->PaymentMeansCode = 42;
+      $altinn_invoice->PaymentMeans->PaymentDueDate = $due_date;
+      $altinn_invoice->PaymentMeans->PayeeFinancialAccount->ID = $bank_account_number;
+      $altinn_invoice->PaymentMeans->PayeeFinancialAccount->Name = 'Bank';
+      // Missing info! No way to determine which of the accounts is the bank account used for tax.
```
Comment (mladjo2505) [inline modules/invoice/model/invoice.class.php]: Which bank account to choose?

File: modules/invoice/model/invoice.class.php [hunk 7]
```diff
@@ -1326,5 +1328,143 @@ function CheckIfAnythingChanged($InvoiceID, $InvoiceLineIDs, &$args) {
...
+      // $altinn_invoice->PaymentMeans->PayerFinancialAccount->Name  = 'Bank';
+
+      $altinn_invoice->PaymentMeans->InstructionID = $kid;
+      $altinn_invoice->PaymentMeans->InstructionNote = "KID";
+
+      // Invoice lines
+      $taxH[0]->TaxableAmount = $amount;
+      $taxH[0]->TaxAmount = 0;
+
+      $altinn_invoice->InvoiceLine[0]->ID = 0;
```
Comment (mladjo2505) [inline modules/invoice/model/invoice.class.php:1424]: Maybe not necessary

File: modules/invoice/model/invoice.class.php [hunk 8]
```diff
@@ -1326,5 +1328,143 @@ function CheckIfAnythingChanged($InvoiceID, $InvoiceLineIDs, &$args) {
...
+      $altinn_invoice->InvoiceLine[0]->LineExtensionAmount = $amount;
+      $altinn_invoice->InvoiceLine[0]->TaxTotal->TaxAmount = 0;
+      $altinn_invoice->InvoiceLine[0]->TaxTotal->TaxSubtotal->TaxableAmount = $amount;
+      $altinn_invoice->InvoiceLine[0]->TaxTotal->TaxSubtotal->TaxAmount = 0;
+      $altinn_invoice->InvoiceLine[0]->TaxTotal->TaxSubtotal->Percent = 0;
+      $altinn_invoice->InvoiceLine[0]->TaxTotal->TaxSubtotal->TaxCategory->TaxScheme->ID = 'VAT';
+
+      $altinn_invoice->InvoiceLine[0]->Item->Name = $invoice_type . " TAX";
+      $altinn_invoice->InvoiceLine[0]->Item->Description = $invoice_type . " TAX for " . $periode;
```
Comment (mladjo2505) [inline modules/invoice/model/invoice.class.php]: Set `$periode` variable. Maybe from `Mottak->kalendermaaned`

File: modules/invoice/model/invoice.class.php [hunk 9]
```diff
@@ -1326,5 +1328,143 @@ function CheckIfAnythingChanged($InvoiceID, $InvoiceLineIDs, &$args) {
...
+
+      $altinn_invoice->InvoiceLine[0]->Price->PriceAmount = $amount;
+      $altinn_invoice->InvoiceLine[0]->Price->BaseQuantity = 1;
+
+      $altinn_invoice->TaxTotal['TaxAmount'] = 0;
+
+      foreach($taxH as $VatPercent => $vat) {
```
Comment (mladjo2505) [inline modules/invoice/model/invoice.class.php:1439]: Maybe directly hardcode the values from params here since we only have for 0% tax group.

**Comment by stoivo** [comment]
- [ ] Publishing date is last sent date(What is used for issue date now). [Unable to send from LODO]
- [x] Issue date should be first date in the period. 
- [x] Should use the NO:ORGNR of parrent company for customer.
- [x] Keep 0% tax lines
- [x] Have note (AGA/FTR 2016-07) on invoice itself and on line description but _not_ on line additional info.

File: modules/invoice/model/invoice.class.php (hunk 6 above)
Comment (stoivo) [inline modules/invoice/model/invoice.class.php]: Try not send it, I think it is selected in fb.

File: modules/invoice/model/invoice.class.php (hunk 4 above)
Comment (mladjo2505) [inline modules/invoice/model/invoice.class.php]: @stoivo This is for you to think about and decide the best way to do it.

**Comment by mladjo2505** [comment]
There are still some unresolved issues about thing that we are unable to send:
Supplier address and contact info and bank account for the customer of the invoice.

**Comment by stoivo** [comment]
To get data for all kommuner use rake task in fakturabank.
Add that to CSV.
Add those columns to kommune table, but don't display it anywhere but the list(dropdown).

**Comment by mladjo2505** [comment]
@stoivo After talking with @hildegunn95 she noted that we do have the company's bank account in the accounting information tab for company settings. She thinks we can use this as the bank account for the customer of the AGA/FTR invoice we send to fakturaBank. Do you agree?

**Comment by stoivo** [comment]
Yes.

**Comment by mladjo2505** [comment]
@stoivo What should we do with the rake task on fakturaBank used to update the kommunes csv?
Nikola and me had the idea to implement this as a cron job(each 6 months or so) on fakturaBank and have an action in LODO to update it, or do it manually?

File: modules/invoice/model/invoice.class.php [hunk 10]
```diff
@@ -1326,5 +1328,128 @@ function CheckIfAnythingChanged($InvoiceID, $InvoiceLineIDs, &$args) {
...
+      } else {
+        $xml = simplexml_load_string($file_contents);
+      }
+
+      $altinn_invoice = new stdClass();
+      $taxH = array();
+
+      $altinn_reference = $altinn_report4_row->res_ArchiveReference;
+      $recieved_messages = $xml->Mottak->mottattLeveranse;
+      // get the issue date from recieved messages
```
Comment (mladjo2505) [inline modules/invoice/model/invoice.class.php]: remove unnecessary comment

File: modules/invoice/model/invoice.class.php [hunk 11]
```diff
@@ -1326,5 +1328,128 @@ function CheckIfAnythingChanged($InvoiceID, $InvoiceLineIDs, &$args) {
...
+      $altinn_invoice->IssueDate = $issue_date;
+      $altinn_invoice->Note = $invoice_type . ' ' . strftime('%Y-%m', strtotime($issue_date));
+      $altinn_invoice->DocumentCurrencyCode = exchange::getLocalCurrency();
+
+      // Invoice supplier
+      $altinn_invoice->AccountingSupplierParty->Party->PartyLegalEntity->CompanyID = $kommune_orgno;
+
+      $altinn_invoice->AccountingSupplierParty->Party->PartyName->Name = $kommune_name;
+
+      $altinn_invoice->AccountingSupplierParty->Party->PostalAddress->StreetName = trim($kommune->Address1 . ' ' . $kommune->Address2 . ' ' . $kommune->Address3);
```
Comment (mladjo2505) [inline modules/invoice/model/invoice.class.php]: `'\n'` instead of `' '`

File: modules/invoice/model/invoice.class.php (hunk 5 above)
Comment (mladjo2505) [inline modules/invoice/model/invoice.class.php]: Change comment to `// Company address and contact info`

File: modules/kommune/model/kommune.class.php [hunk 12]
```diff
@@ -12,6 +12,9 @@ class kommune {
   private $kommunes_in_db_by_number = array();
   private $kommune_data_for_select = array();
 
+  // a list of columns allowed to search by
+  public $permitted_columns = array('KommuneID', 'KommuneNumber', 'KommuneName', 'County', 'Sone', 'BankAccountNumber', 'OrgNumber', 'OrgName', 'OrganisationForm', 'Comments');
```
Comment (mladjo2505) [inline modules/kommune/model/kommune.class.php]: Add `Telephone, Telefax, Email, Mobile, Webpage, Address1, Address2, Address3, ZipCode, City`

**Comment by mladjo2505** [comment]
Show new CSV columns only in dropdown or only not in dropdown?

**EDIT**: Only do not show in dropdown

What if we send to altinn for more kommunes?

**EDIT**: All good, paid to one kommune, then it gets redistributed afterwards by that kommune.

What to do when kommune doesn’t have bank account, should we use the kommune linked to parent company?

**EDIT**: If kommune has no bank account show the warning message "Bank account not found in kommune list, contact the administrator before you can send to fakturaBank"

**Comment by sls-lnluksa** [comment]
First round review: Nikola Lukovic (sls-lnluksa)
Final review with: @stoivo

**Comment by mladjo2505** [comment]
- [x] If kommune has no bank account show the warning message "Bank account not found in kommune list, contact the administrator before you can send to fakturaBank"
- [x] Issue date is the last date in period

File: modules/invoice/model/invoice.class.php [hunk 13]
```diff
@@ -1327,5 +1329,126 @@ function CheckIfAnythingChanged($InvoiceID, $InvoiceLineIDs, &$args) {
...
+        return false; // File can't be read
+      } else {
+        $xml = simplexml_load_string($file_contents);
+      }
+
+      $altinn_invoice = new stdClass();
+      $taxH = array();
+
+      $altinn_reference = $altinn_report4_row->res_ArchiveReference;
+      $recieved_messages = $xml->Mottak->mottattLeveranse;
```
Comment (mladjo2505) [inline modules/invoice/model/invoice.class.php]: unused, remove

File: modules/invoice/model/invoice.class.php [hunk 14]
```diff
@@ -1327,5 +1329,126 @@ function CheckIfAnythingChanged($InvoiceID, $InvoiceLineIDs, &$args) {
...
+      $issue_date = (string) $xml->Mottak->kalendermaaned . '-01';
```
Comment (mladjo2505) [inline modules/invoice/model/invoice.class.php]: Should be the last day of the month, not the first

**Pull request merged**